### ISSUE TITLE
Modify global scope variables from Thread context not possible?

### ISSUE THREAD

**Issue by zelgerj** [krakjoe/pthreads#426]
Hi there,

i just wondered if it's not possible to write back values to global scope variables using globally or if it's still a bug / feature.

```php
<?php

class GlobalExecutor extends \Thread {
    public function __construct($callable) {
        $this->callable = $callable;
        $this->start(PTHREADS_INHERIT_ALL | PTHREADS_ALLOW_GLOBALS);
        $this->join();
    }
    public function run() {
        \Thread::globally($this->callable);
    }
}

$testVar = 'before';

$ex = new GlobalExecutor(function() {
    $testVar = 'after';
});

var_dump($testVar);
```

The output should be...
```
string(5) "after"
```

...instead of:
```
string(5) "before"
```

Or am i totally wrong here?

Cheers!
Johann

**Comment by staabm** [comment]
In regular PHP you cannot override the value of a global var using the notation you are using

```php
$testVar = 'before';

function a() {
    $testVar = 'after';
};

a();
var_dump($testVar);
```

you either need to use `$GLOBALS` 
```php
$testVar = 'before';

function a() {
   // alternatively:  $GLOBALS['testVar'] = 'after';
    global $testVar;
    $testVar = 'after';
};

a();
var_dump($testVar);
```

or `use` the var by reference
```php
$testVar = 'before';

function a() use (&$testVar) {
    $testVar = 'after';
};

a();
var_dump($testVar);
```

It sounds like a bug in your code instead of pthread.

**Comment by zelgerj** [comment]
Hi, thanks for that response. But all your example code doesn't use threading functionality at all... I'am quite sure that php itself works well without pthreads...:)

Please check it out with threading functionality...

If i use the var by reference:
```php
<?php

class GlobalExecutor extends \Thread {
    public function __construct($callable) {
        $this->callable = $callable;
        $this->start(PTHREADS_INHERIT_ALL | PTHREADS_ALLOW_GLOBALS);
        $this->join();
    }
    public function run() {
        \Thread::globally($this->callable);
    }
}

$testVar = 'before';

var_dump($testVar);

$ex = new GlobalExecutor(function() use(&$testVar) {
    $testVar = "after";
});

var_dump($testVar);
```

The result seems to be a bit buggy :)
```
string(6) "before"
string(5) "
```

should be
```
string(6) "before"
string(5) "after"
```

**Comment by SirSnyder** [comment]
The result is right. GlobalExecutor runs in its own context. To share variables across context boundaries you will have to use Threaded

```php
<?php

class GlobalExecutor extends \Thread {
    public function __construct() {
        $this->start(PTHREADS_INHERIT_ALL | PTHREADS_ALLOW_GLOBALS);
        $this->join();
    }
    public function run() {
        global $shared;
        $shared['testVar'] = "after";
    }
}

global $shared;
$shared = new Threaded();

$shared['testVar'] = 'before';

var_dump($shared['testVar']);

$ex = new GlobalExecutor();

var_dump($shared['testVar']);
```

**Comment by zelgerj** [comment]
Thanks for reply @SirSnyder. Your example is correct... but not what to do... you are not using the `Thread::globally` functionality... Joe says in examples that the code will be executed in global scope itself, so no context sharing should be needed at all...

**Comment by SirSnyder** [comment]
Maybe you have to define $testVar explicit as global in your example? Like

```php
global $testVar;
$testVar = 'before';
```
But my experience is, that the global context just gets copied to a new thread. There is no sharing. Therefore I use one instance of Threaded in the global context, which gets copied to new threads and can be accessed.

Sharing a closure between contexts could be become another problem.

**Comment by zelgerj** [comment]
Thanks for that advice but nothing changes.

I know that the global context will be copied to a new thread but i think ´Thread::globally´ should work vice versa. If you have a look at the globally implementation the closure will be executed at global scope... otherwise that function would not make sense at all...

@krakjoe Could you give us a statement to clear it up? Thanks :)

Cheers,
Johann.

**Comment by chibisuke** [comment]
you're mixing two features of pthreads that seem to cause a lot of confusion. 
I found the original isue thread where those features were discussed and implemented #245 

Threaded::globaly executes a closure using the symbol table of the global scope as its local symbol table. Meaning all global variables will behave as locals inside the closure.
So as far as this goes the code of the OP is correct.

however I'm not 100% sure on the exact workings of PTHREADS_ALLOW_GLOBALS, but I guess its not possible to gain enougth control over those hashtables to make them share data directly. I may be mistaken though.

However what should be possible is, that you create a pthreads objects (Threaded/Stackable) and put it in the global scope before starting the thread. And then use the reference to this object to share data by putting them inside.

**Comment by krakjoe** [comment]
Closing, you seem to have found your way ...
